components/provider_cldr/src/transform/likelysubtags.rs

```diff
@@ -0,0 +1,148 @@
+// This file is part of ICU4X. For terms of use, please see the file
+// called LICENSE at the top level of the ICU4X source tree
+// (online at: https://github.com/unicode-org/icu4x/blob/master/LICENSE ).
+use crate::error::Error;
+use crate::reader::open_reader;
+use crate::CldrPaths;
+use icu_provider::prelude::*;
+use icu_provider::structs::likelysubtags::*;
+use std::borrow::Cow;
+use std::convert::TryFrom;
+use std::marker::PhantomData;
+
+/// All keys that this module is able to produce.
+pub const ALL_KEYS: [ResourceKey; 1] = [key::LIKELY_SUBTAGS_V1];
+
+/// A data provider reading from CLDR JSON likely subtags rule files.
+#[derive(PartialEq, Debug)]
+pub struct LikelySubtagsProvider<'d> {
+    data: cldr_json::Resource,
+    _phantom: PhantomData<&'d ()>, // placeholder for when we need the lifetime param
+}
+
+impl TryFrom<&dyn CldrPaths> for LikelySubtagsProvider<'_> {
+    type Error = Error;
+    fn try_from(cldr_paths: &dyn CldrPaths) -> Result<Self, Self::Error> {
+        let data: cldr_json::Resource = {
+            let path = cldr_paths
+                .cldr_core()?
+                .join("supplemental")
+                .join("likelySubtags.json");
+            serde_json::from_reader(open_reader(&path)?).map_err(|e| (e, path))?
+        };
+        Ok(Self {
+            data,
+            _phantom: PhantomData,
+        })
+    }
+}
+
+impl<'d> TryFrom<&'d str> for LikelySubtagsProvider<'d> {
+    type Error = serde_json::error::Error;
+    /// Attempt to parse a JSON string.
+    fn try_from(s: &'d str) -> Result<Self, Self::Error> {
+        let data: cldr_json::Resource = serde_json::from_str(s)?;
+        Ok(Self {
+            data,
+            _phantom: PhantomData,
+        })
+    }
+}
+
+impl<'d> KeyedDataProvider for LikelySubtagsProvider<'d> {
+    fn supports_key(resc_key: &ResourceKey) -> Result<(), DataError> {
+        if resc_key.category != ResourceCategory::LikelySubtags {
+            return Err((&resc_key.category).into());
+        }
+        if resc_key.version != 1 {
+            return Err(resc_key.into());
+        }
+        Ok(())
+    }
+}
+
+impl<'d> DataProvider<'d, LikelySubtagsV1> for LikelySubtagsProvider<'d> {
+    fn load_payload(
+        &self,
+        req: &DataRequest,
+    ) -> Result<DataResponse<'d, LikelySubtagsV1>, DataError> {
+        LikelySubtagsProvider::supports_key(&req.resource_path.key)?;
+        let langid = req.resource_path.options.langid.clone();
```

Comment:
I think you shouldn't clone until you actually need to clone.  You don't need the clone until line 77.